File: swift/Sources/benchmarks/GraphPath.swift
import Collections
import Foundation

class GraphPathBenchmark: BenchmarkProtocol {
  final class Graph {
    let vertices: Int
    let jumps: Int
    let jumpLen: Int
    var adj: [[Int]]

    init(vertices: Int, jumps: Int = 3, jumpLen: Int = 100) {
      self.vertices = vertices
      self.jumps = jumps
      self.jumpLen = jumpLen
      self.adj = Array(repeating: [Int](), count: vertices)
    }

    func addEdge(_ u: Int, _ v: Int) {
      adj[u].append(v)
      adj[v].append(u)
    }

    func generateRandom() {

      for i in 1..<vertices {
        addEdge(i, i - 1)
      }

      for v in 0..<vertices {
        let numJumps = Helper.nextInt(max: jumps)
        for _ in 0..<numJumps {
          let offset = Helper.nextInt(max: jumpLen) - jumpLen / 2
          let u = v + offset

          if u >= 0 && u < vertices && u != v {
            addEdge(v, u)
          }
        }
      }
    }
  }

  var graph: Graph!
  private var resultVal: UInt32 = 0

  init() {}

  func prepare() {
    let vertices = Int(configValue("vertices") ?? 0)
    let jumps = Int(configValue("jumps") ?? 0)
    let jumpLen = Int(configValue("jump_len") ?? 0)

    graph = Graph(vertices: vertices, jumps: jumps, jumpLen: jumpLen)
    graph.generateRandom()
  }

  func test() -> Int64 {
    return 0
  }

  func run(iterationId: Int) {
    resultVal &+= UInt32(test())
  }

  var checksum: UInt32 {
    return resultVal
  }

  func name() -> String {
    return ""
  }
}

final class GraphPathBFS: GraphPathBenchmark {
  override init() {
    super.init()
  }

  private func bfsShortestPath(_ start: Int, _ target: Int) -> Int {
    if start == target { return 0 }

    var visited = [Bool](repeating: false, count: graph.vertices)
    var queue: [(Int, Int)] = []
    queue.reserveCapacity(graph.vertices)
    visited[start] = true
    queue.append((start, 0))
    var front = 0

    while front < queue.count {
      let (v, dist) = queue[front]
      front += 1

      for neighbor in graph.adj[v] {
        if neighbor == target { return dist + 1 }
        if !visited[neighbor] {
          visited[neighbor] = true
          queue.append((neighbor, dist + 1))
        }
      }
    }
    return -1
  }

  override func test() -> Int64 {
    return Int64(bfsShortestPath(0, graph.vertices - 1))
  }

  override func name() -> String {
    return "Graph::BFS"
  }
}

final class GraphPathDFS: GraphPathBenchmark {
  override init() {
    super.init()
  }

  private func dfsFindPath(_ start: Int, _ target: Int) -> Int {
    if start == target { return 0 }

    var visited = [Bool](repeating: false, count: graph.vertices)
    var stack: [(Int, Int)] = [(start, 0)]
    var bestPath = Int.max

    while !stack.isEmpty {
      let (v, dist) = stack.removeLast()

      if visited[v] || dist >= bestPath {
        continue
      }

      visited[v] = true

      for neighbor in graph.adj[v] {
        if neighbor == target {
          if dist + 1 < bestPath {
            bestPath = dist + 1
          }
        } else if !visited[neighbor] {
          stack.append((neighbor, dist + 1))
        }
      }
    }

    return bestPath == Int.max ? -1 : bestPath
  }

  override func test() -> Int64 {
    return Int64(dfsFindPath(0, graph.vertices - 1))
  }

  override func name() -> String {
    return "Graph::DFS"
  }
}

struct GraphAStarEntry: Comparable {
  let priority: Int
  let vertex: Int

  static func < (lhs: GraphAStarEntry, rhs: GraphAStarEntry) -> Bool {
    if lhs.priority != rhs.priority {
      return lhs.priority < rhs.priority
    }
    return lhs.vertex < rhs.vertex
  }

  static func == (lhs: GraphAStarEntry, rhs: GraphAStarEntry) -> Bool {
    lhs.priority == rhs.priority && lhs.vertex == rhs.vertex
  }
}

final class GraphPathAStar: GraphPathBenchmark {
  override init() {
    super.init()
  }

  private func heuristic(_ v: Int, _ target: Int) -> Int {
    return target - v
  }

  private func aStarShortestPath(_ start: Int, _ target: Int) -> Int {
    if start == target { return 0 }

    let n = graph.vertices
    var gScore = [Int](repeating: Int.max, count: n)
    var bestF = [Int](repeating: Int.max, count: n)

    gScore[start] = 0
    let fStart = heuristic(start, target)
    bestF[start] = fStart

    var openSet = Heap<GraphAStarEntry>(minimumCapacity: n)
    openSet.insert(GraphAStarEntry(priority: fStart, vertex: start))

    while !openSet.isEmpty {
      guard let entry = openSet.popMin() else { break }
      let current = entry.vertex

      if current == target {
        return gScore[current]
      }

      for neighbor in graph.adj[current] {
        let tentativeG = gScore[current] + 1

        if tentativeG < gScore[neighbor] {
          gScore[neighbor] = tentativeG
          let fNew = tentativeG + heuristic(neighbor, target)

          if fNew < bestF[neighbor] {
            bestF[neighbor] = fNew
            openSet.insert(GraphAStarEntry(priority: fNew, vertex: neighbor))
          }
        }
      }
    }

    return -1
  }

  override func test() -> Int64 {
    return Int64(aStarShortestPath(0, graph.vertices - 1))
  }

  override func name() -> String {
    return "Graph::AStar"
  }
}

File: cpp/src/graph_path.cpp
#include "graph_path.hpp"
#include <climits>
#include <queue>
#include <stack>

GraphPathBenchmark::Graph::Graph(int vertices, int jumps, int jump_len)
    : vertices(vertices), jumps(jumps), jump_len(jump_len), adj(vertices) {}

void GraphPathBenchmark::Graph::add_edge(int u, int v) {
  adj[u].push_back(v);
  adj[v].push_back(u);
}

void GraphPathBenchmark::Graph::generate_random() {
  for (int i = 1; i < vertices; i++) {
    add_edge(i, i - 1);
  }

  for (int v = 0; v < vertices; v++) {
    int num_jumps = Helper::next_int(jumps);
    for (int j = 0; j < num_jumps; j++) {
      int offset = Helper::next_int(jump_len) - jump_len / 2;
      int u = v + offset;

      if (u >= 0 && u < vertices && u != v) {
        add_edge(v, u);
      }
    }
  }
}

GraphPathBenchmark::GraphPathBenchmark() : result_val(0) {}

void GraphPathBenchmark::prepare() {
  int vertices = static_cast<int>(config_val("vertices"));
  int jumps = static_cast<int>(config_val("jumps"));
  int jump_len = static_cast<int>(config_val("jump_len"));
  graph = std::make_unique<Graph>(vertices, jumps, jump_len);
  graph->generate_random();
}

void GraphPathBenchmark::run(int iteration_id) {
  (void)iteration_id;
  result_val += test();
}

uint32_t GraphPathBenchmark::checksum() { return result_val; }

std::string GraphPathBFS::name() const { return "Graph::BFS"; }

int GraphPathBFS::bfs_shortest_path(int start, int target) {
  if (start == target)
    return 0;

  std::vector<uint8_t> visited(graph->vertices, 0);
  std::queue<std::pair<int, int>> queue;

  visited[start] = 1;
  queue.push({start, 0});

  while (!queue.empty()) {
    auto [v, dist] = queue.front();
    queue.pop();

    for (int neighbor : graph->adj[v]) {
      if (neighbor == target)
        return dist + 1;
      if (visited[neighbor] == 0) {
        visited[neighbor] = 1;
        queue.push({neighbor, dist + 1});
      }
    }
  }
  return -1;
}

int64_t GraphPathBFS::test() {
  return bfs_shortest_path(0, graph->vertices - 1);
}

std::string GraphPathDFS::name() const { return "Graph::DFS"; }

int GraphPathDFS::dfs_shortest_path(int start, int target) {
  if (start == target)
    return 0;

  std::vector<uint8_t> visited(graph->vertices, 0);
  std::stack<std::pair<int, int>> stack;
  int best_path = INT_MAX;

  stack.push({start, 0});

  while (!stack.empty()) {
    auto [v, dist] = stack.top();
    stack.pop();

    if (visited[v] == 1 || dist >= best_path)
      continue;
    visited[v] = 1;

    for (int neighbor : graph->adj[v]) {
      if (neighbor == target) {
        if (dist + 1 < best_path)
          best_path = dist + 1;
      } else if (visited[neighbor] == 0) {
        stack.push({neighbor, dist + 1});
      }
    }
  }
  return (best_path == INT_MAX) ? -1 : best_path;
}

int64_t GraphPathDFS::test() {
  return dfs_shortest_path(0, graph->vertices - 1);
}

std::string GraphPathAStar::name() const { return "Graph::AStar"; }

int GraphPathAStar::astar_shortest_path(int start, int target) {
  if (start == target)
    return 0;

  int n = graph->vertices;

  std::vector<int> g_score(n, INT_MAX);
  std::vector<int> best_f(n, INT_MAX);

  g_score[start] = 0;
  int f_start = heuristic(start, target);
  best_f[start] = f_start;

  using QueueType =
      std::priority_queue<Node, std::vector<Node>, std::greater<Node>>;
  QueueType open_set;
  open_set.push({start, f_start});

  while (!open_set.empty()) {
    Node current = open_set.top();
    open_set.pop();

    int current_vertex = current.vertex;

    if (current_vertex == target)
      return g_score[current_vertex];

    for (int neighbor : graph->adj[current_vertex]) {
      int tentative_g = g_score[current_vertex] + 1;

      if (tentative_g < g_score[neighbor]) {
        g_score[neighbor] = tentative_g;
        int f_new = tentative_g + heuristic(neighbor, target);

        if (f_new < best_f[neighbor]) {
          best_f[neighbor] = f_new;
          open_set.push({neighbor, f_new});
        }
      }
    }
  }

  return -1;
}

int64_t GraphPathAStar::test() {
  return astar_shortest_path(0, graph->vertices - 1);
}